Cache component dependencies separately for each service

Legioth opened the pull request:
It's in theory not necessary to cache separately for each service since
any service (of the same type) should produce the same result. What is
necessary is to avoid caching the dummy result produced when there is no
service at all.

Also adds protection in ReflectionCache to avoid accidentally relying on
any current instance when computing cache values.
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/flow/3967)
<!-- Reviewable:end -->

denis-anisimov commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3967#-:-LAm1uGZ-5HTTMm4GAok:bnfp4nl)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
There are conflicts that needs to be resolved.

Legioth commented:
Fixed conflicts and changed this PR to primarily fix the case that was accidentally relying on a current instance when computing a cached value.

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3967#-:-LAo2OgeAa6WuPZZ8SRJ:bkwtsdi)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[flow-server/src/test/java/com/vaadin/flow/internal/ReflectionCacheTest.java, line 97 at r2](https://reviewable.io/reviews/vaadin/flow/3967#-LAqEYex2EqZM43HjP0I:-LAqEYey0PFSzh0KNnNO:bab2exp) ([raw file](https://github.com/vaadin/flow/blob/f121208a1b68213e30eec82cfc13138de44e538a/flow-server/src/test/java/com/vaadin/flow/internal/ReflectionCacheTest.java#L97)):*
> ```
>     public void testCurrentInstances() {
> ```

Method name doesn't really say what it tests.
An no need `test` as a prefix. It also doesn't help anyhow.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/3967)*
<!-- Sent from Reviewable.io -->

Legioth commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[flow-server/src/test/java/com/vaadin/flow/internal/ReflectionCacheTest.java, line 97 at r2](https://reviewable.io/reviews/vaadin/flow/3967#-LAqEYex2EqZM43HjP0I:-LAqb5Th2t160hKbMAqB:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/f121208a1b68213e30eec82cfc13138de44e538a/flow-server/src/test/java/com/vaadin/flow/internal/ReflectionCacheTest.java#L97)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
>     public void testCurrentInstances() {
> ```

Method name doesn't really say what it tests.
An no need `test` as a prefix. It also doesn't help anyhow.

</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/3967)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/3967#-:-LAqcaaV1nF-t-rXPQV3:bnfp4nl)*
<!-- Sent from Reviewable.io -->
